### DISCUSSION TITLE
Solver stuck when running with high number of CPUs

### DISCUSSION THREAD

**Discussion by mojtabaAmiraslanpour** [UvaCsl/HemoCell#57] [General discussion]
Hi,

I have this problem that the solver is stuck without any errors mainly in iterations less than 100K when I am running my case with a higher number of CPU cores. Am I doing something wrong? So for instance this case is stuck at iter = 14K. I have run the same case without any problems with 80 CPU cores. 

```
Size of the multi-block:     1026-by-253-by-66
Number of atomic-blocks:     256
Smallest atomic-block:       128-by-31-by-16
Largest atomic-block:        129-by-32-by-17
Number of allocated cells:   17.1321 million
Fraction of allocated domain: 100 percent
(SanityCheck) one of the dimensions of the largest atomic block is more than 25.
  This is inefficient, The best performance is with 16x16x16 blocks.
  It is recommended to adjust the number of processors or the sparseBlockStructure accordingly.
```

I am not sure if this is because of having too thin atomic-blocks in the domain. Does anybody have any clues?

Thank you

**Comment by gzavo** [comment]
Hi @mojtabaAmiraslanpour,

I did see similar problems when the atomic block sizes were very uneven, albeit at much larger deployments (>10k cores).
It triggered some kind of a race condition in combination with specific MPI versions. If the same problem runs fine with a different atomic block structure, then I think it is a strong clue.

The next HemoCell version will include a new xml tags (mABx, mABy, mABz) that will allow to define how many divisions you want along the global axis, so you will be able to influence the atomic block shapes. This release is ready for a while now, so I hope it will appear in a few days on github. For now, changing the number of cores (even by 1 or 2) might provide a (not too elegant) solution.

Bests,
Gabor

**Comment by mojtabaAmiraslanpour** [comment]
@gzavo 

That is great. I am looking forward to it. How about load balancing with ParMETIS? Do you think using it will alleviate the problem?
I am not using ParMETIS for now.

**Reply by gzavo** [reply]
I don't think it will help. Load-balancing is designed to redistribute the simulation during run, but uses the same method to redefine the atomic block structure as in the upcoming version. If all goes well it will appear on Monday.
